torchqdynamics/odeint.py

```diff
@@ -1,80 +1,124 @@
-        return _fixed_odeint(qsolver, y0, tsave)
+        return _fixed_odeint(
+            qsolver, y0, t_save, qsolver.options.dt, exp_ops, save_states
+        )
     elif isinstance(qsolver.options, AdaptativeStep):
-        return _adaptive_odeint(qsolver, y0, tsave)
+        return _adaptive_odeint(qsolver, y0, t_save, exp_ops, save_states)
 
 
-def odeint_inplace(qsolver, y0, tsave):
+# For now we use *args and **kwargs for helper methods that are not implemented to ease the potential API
+# changes that could occur later. When a method is implemented the methods should take the same arguments
+# as all others
+def _odeint_inplace(*args, **kwargs):
     # TODO: Simple solution for now so torch does not store gradients. This
     #       is probably slower than a genuine in-place solver.
     with torch.no_grad():
-        return odeint_main(qsolver, y0, tsave)
+        return _odeint_main(*args, **kwargs)
 
 
-def odeint_adjoint(qsolver, y0, tsave, parameters):
+def _odeint_adjoint(*_args, **_kwargs):
     raise NotImplementedError
 
 
-def _adaptive_odeint(qsolver, y0, tsave):
+def _adaptive_odeint(*_args, **_kwargs):
     raise NotImplementedError
 
 
-def _fixed_odeint(qsolver, y0, tsave):
-    # initialize save tensor
-    ysave = torch.zeros(len(tsave), *y0.shape).to(y0)
+def _fixed_odeint(qsolver, y0, t_save, dt, exp_ops, save_states):
+    # Initialize save tensor
+    y_save = None
+    if save_states:
+        y_save = torch.zeros(len(t_save), *y0.shape).to(y0)
+
+    if len(exp_ops) > 0:
+        exp_save = torch.zeros(len(t_save), len(exp_ops)).to(
+            device=y0.get_device(), dtype=torch.float
+        )
+    else:
+        exp_save = None
+
+    # Save first step
     save_counter = 0
-    if tsave[0] == 0.0:
-        ysave[0] = y0
+    if t_save[0] == 0.0:
+        if save_states:
+            y_save[0] = y0
+        for j, op in enumerate(exp_ops):
+            exp_save[save_counter, j] = torch.trace(op @ y0)
         save_counter += 1
 
-    # get qsolver fixed time step
-    dt = qsolver.options.dt
-
-    # run the ODE routine
-    t, y = 0.0, y0
-    while t < tsave[-1]:
-        # iterate solution
+    # Run the ODE routine
+    y = y0
+    t, t_max = 0, max(t_save)
+    while t < t_max:
+        # Iterate solution
         y = qsolver.forward(t, dt, y)
-        t = t + dt
 
-        # save solution
-        if t >= tsave[save_counter]:
-            ysave[save_counter] = y
+        # Save solution
+        if t >= t_save[save_counter]:
+            if save_states:
+                y_save[save_counter] = y
+
+            for j, op in enumerate(exp_ops):
+                exp_save[save_counter, j] = torch.trace(op @ y)
             save_counter += 1
 
-    return ysave
+    return y_save, exp_save
 
 
-def check_tsave(tsave):
-    """Check that `tsave` is valid (it must be a non-empty 1D tensor sorted in
+def check_t(t):
+    """Check that `t_save` is valid (it must be a non-empty 1D tensor sorted in
```

Comment:
We should either make the method generic (i.e. it checks a generic time tensor), and never refer to `tsave` within the method, or explicitly keep the name `check_tsave` IMO.